design_notes/XDN05 - xlang Binary Interface.md

```diff
@@ -12,6 +12,233 @@ status: draft
+enums can be additively versioned over time, it is possible to add new ones over time. Since
+supporting a given information category is optional, adding new ones would not break existing
+objects. Furthermore, it reduces the size of virtual function table that every xlang object needs
+to support.
+
+Over time, we expect the list of object information categories to grow. Unlike COM interfaces, we
+expect the list of object information categories to be managed centrally by the xlang project.
+
+#### String Representation
+
+Most mainstream languages including
+[JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toString),
+[Python](https://docs.python.org/3/library/functions.html#func-str),
+[.NET](https://docs.microsoft.com/en-us/dotnet/api/system.object.tostring?view=netframework-4.7.2),
+[Java](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html#toString()) and
+[Objective-C](https://developer.apple.com/documentation/objectivec/1418956-nsobject/1418746-description)
+have a mechanism to retrieve the string representation of an object. 
+
+The StringRepresentation information category retrieves the string representation of the object in
+question. The property type for StringRepresentation is an xlang_string.
+
+While all information categories are optional, it is highly recommended that all objects support
+StringRepresentation. By default, this method can return the type's name, however it can be
+implemented in whatever manner makes most sense for the underlying object.
+
+#### Type Information
+
+Dynamic languages such as JavaScript and Python often need to be able to discover information
+about an object at runtime. In order to do this, xlang object must have a mechanism to find the
+metadata for arbitrary xlang objects.
+
+The TypeName information category retrieves the fully namespace qualified name of the type in
+question. The property type for TypeName is an xlang_string.
+
+While all information categories are optional, it is highly recommended that all objects support
+TypeName in order to be usable from dynamic languages.
+
+#### Memory Usage
+
+In languages C# and Java, garbage collection happens on background threads on a non-deterministic
+schedule. These garbage collectors typically depend on an omniscient understanding of memory usage.
+The memory used by xlang components is not typically visible to these garbage collectors, leading
+to uninformed decision making regarding collection scheduling. If these systems could retrieve
+the memory usage of arbitrary xlang objects, their scheduling heuristics would be better and the
+system would run more smoothly overall.
+
+The MemoryUsage information category retrieves the memory usage in bytes of the object in question.
+The property type for MemoryUsage is an unsigned 32-bit integer.
+
+## Weak References
+
+One issue faced in COM based systems is circular references - i.e. two objects who each hold a
```

Comment:
Not just COM. Any reference-counting model. e.g. std::shared_ptr.

Reply:
You're right, but is that really imporant to include here?